### PR TITLE
[DotNetCoreCLIV2]handling semver v2 runtime preview versions

### PR DESCRIPTION
**Pull request by tejasd1990**
Fix for issue https://github.com/microsoft/azure-pipelines-tasks/issues/11612

### CODE REVIEW

**Comment by hiyadav** [comment]
@tejasd1990 can you run some tests while downloading 
1. sdk
2. runtime 

of semver version 1.0 and 2.0 

with

1. exact versioning: 2.2.105
2. 3.0.x / 3.x  (with include preview enabled and disabled)


We just need to be sure that we aren't breaking anything here.

**Comment by tejasd1990** [comment]
@hiyadav I have already tested various scenarios with the new change including versions which are v1 and v2

File: Tasks/UseDotNetV2/models.ts [hunk 1]
```diff
@@ -113,34 +115,6 @@ export class VersionParts {
         }
     }
 
-    /**
-     * Validate the version if this string is a explicit version number. Returns an exception if the version number is not explicit.
-     * @param version the input version number as string
-     */
-    private static ValidateExplicitVersionNumber(version: string): void {
```
Comment (bishal-pdMSFT) [inline Tasks/UseDotNetV2/models.ts]: Why are you removing this functionality. it checks whether user has entered an explicit version or not

File: Tasks/UseDotNetV2/models.ts [hunk 2]
```diff
@@ -121,13 +121,13 @@ export class VersionParts {
...
             // validate version
-            if ((parts.length != 3) || // check if the version has 3 parts
+            if ((parts.length < 3) || // check if the version has at least 3 parts
                 !parts[0] || // The major version must always be set
                 !parts[1] || // The minor version must always be set
                 !parts[2] || // The patch version must always be set
                 Number.isNaN(Number.parseInt(parts[0])) || // the major version number must be a number
                 Number.isNaN(Number.parseInt(parts[1])) || // the minor version number must be a number
-                Number.isNaN(Number.parseInt(parts[2].split('-')[0])) // the patch version number must be a number. (the patch version can have a '-' because of version numbers like: 1.0.0-beta-50)
+                Number.isNaN(Number.parseInt(parts[2].split(/\-|\+/)[0])) // the patch version number must be a number. (the patch version can have a '-', or a '+' because of version numbers like: 1.0.0-beta-50)
```
Comment (bishal-pdMSFT) [inline Tasks/UseDotNetV2/models.ts]: Can you point me to such an example where + is used

File: Tasks/UseDotNetV2/models.ts [hunk 3]
```diff
@@ -149,16 +149,22 @@ export class VersionParts {
...
                 Number.isNaN(Number.parseInt(parts[0])) || // the major version number must be a number
-                ( // The next lines check the minor version
-                    Number.isNaN(Number.parseInt(parts[1])) &&  // the minor version number must be a number
-                    parts[1] != "x" // the minor version can't be `x`
+                (
+                    parts[1] != "x" && // if the minor version is not `x`
+                    (
+                        Number.isNaN(Number.parseInt(parts[1])) || // the minor version number must be a number
+                        (
+                            parts.length > 2 && parts[2] != "x" && // if the patch is not `x`, then its an explicit version
```
Comment (bishal-pdMSFT) [inline Tasks/UseDotNetV2/models.ts]: Will `ValidateVersionSpec` be called if explicit version is specified?

File: Tasks/UseDotNetV2/models.ts (hunk 2 above)
Comment (tejasd1990) [reply Tasks/UseDotNetV2/models.ts]: This is to make it semver compliant. https://semver.org/#backusnaur-form-grammar-for-valid-semver-versions
In Dotnetcore releases I think so far they have not used it, but if they do in the future, then this change will be required.

File: Tasks/UseDotNetV2/models.ts (hunk 3 above)
Comment (tejasd1990) [reply Tasks/UseDotNetV2/models.ts]: ValidateExplicitVersionNumber will be called if global.json is used
ValidateVersionSpec will be called if global.json is not used, irrespective of whether the version is explicit or not.

**Comment by TJEvans** [comment]
This is a slight breaking change as the 'X' version identifier is now case sensitive....

**Comment by tejasd1990** [comment]
Hi @TJEvans , The version identifier has always been case sensitive, it should be lower 'x'. Can you confirm whether it worked for you earlier with capital 'X' as well?

**Comment by TJEvans** [comment]
Prior to the task updating this was the logs, 

2019-11-07T17:23:44.3328779Z ##[section]Starting: Use dotnet sdk 2.2.X
2019-11-07T17:23:44.3759718Z ==============================================================================
2019-11-07T17:23:44.3759933Z Task         : Use .NET Core
2019-11-07T17:23:44.3760075Z Description  : Acquires a specific version of the .NET Core SDK from the internet or the local cache and adds it to the PATH. Use this task to change the version of .NET Core used in subsequent tasks. Additionally provides proxy support.
2019-11-07T17:23:44.3760386Z Version      : 2.0.21
2019-11-07T17:23:44.3760508Z Author       : Microsoft Corporation
2019-11-07T17:23:44.3760682Z Help         : https://aka.ms/AA4xgy0
2019-11-07T17:23:44.3760780Z ==============================================================================
2019-11-07T17:23:45.3619447Z Tool to install: .NET Core sdk version 2.2.X.
2019-11-07T17:23:46.4170769Z Found version 2.2.402 in channel 2.2 for user specified version spec: 2.2.X
2019-11-07T17:23:46.4687151Z Version: 2.2.402 was found in cache.
2019-11-07T17:23:46.4736116Z Creating global tool path and pre-pending to PATH.
2019-11-07T17:23:46.4787149Z ##[section]Finishing: Use dotnet sdk 2.2.X

**Pull request merged**